languageserver/integration/codelenses.go

```diff
@@ -159,72 +257,162 @@ func (i *FlowIntegration) entryPointActions(
 	if entryPointInfo.kind == entryPointKindUnknown || entryPointInfo.startPos == nil {
 		return nil, nil
 	}
-	position := *entryPointInfo.startPos
-
-	var title string
-	var argumentListConjunction string
-	var command string
 
-	switch entryPointInfo.kind {
-	case entryPointKindScript:
-		title = "execute"
-		argumentListConjunction = "with"
-		command = CommandExecuteScript
-
-	case entryPointKindTransaction:
-		// Do not show submit button when no active account exists
-		if i.activeAddress == flow.EmptyAddress {
-			return nil, nil
+	position := *entryPointInfo.startPos
+	codelensRange := conversion.ASTToProtocolRange(position, position)
+	var codeLenses []*protocol.CodeLens
+
+	// If emulator is not up, we need to show single codelens proposing to start emulator
+	if i.emulatorState == EmulatorOffline {
+		title := fmt.Sprintf(
+			"%s Emulator is Offline. Click here to start it",
+			prefixOffline,
+		)
+		codeLens := &protocol.CodeLens{
+			Range: codelensRange,
+			Command: &protocol.Command{
+				Title:     title,
+				Command:   ClientStartEmulator,
+				Arguments: []interface{}{uri, "Offline"},
+			},
 		}
+		codeLenses = append(codeLenses, codeLens)
+		return codeLenses, nil
+	}
 
-		title = fmt.Sprintf(
-			"submit with account 0x%s",
-			i.activeAddress.Hex(),
+	// If emulator is not up, we need to provide actionless codelens
+	if i.emulatorState == EmulatorStarting {
+		title := fmt.Sprintf(
+			"%s Emulator is starting up. Please wait \u2026",
+			prefixStarting,
 		)
-		argumentListConjunction = "and"
-		command = CommandSubmitTransaction
+		codeLens := &protocol.CodeLens{
+			Range: codelensRange,
+			Command: &protocol.Command{
+				Title: title,
+			},
+		}
+		codeLenses = append(codeLenses, codeLens)
+		return codeLenses, nil
 	}
 
 	argumentLists := entryPointInfo.pragmaArguments[:]
 
 	// If there are no parameters and no pragma argument declarations,
 	// offer execution using no arguments
-	if len(entryPointInfo.parameters) == 0 && len(argumentLists) == 0 {
+	if len(entryPointInfo.parameters) == 0 {
 		argumentLists = append(argumentLists, []cadence.Value{})
 	}
 
-	codeLenses := make([]*protocol.CodeLens, len(argumentLists))
+	signersList := entryPointInfo.pragmaSignersStrings[:]
+	// TODO: resolve check for amount of signers equals to provided by pragma
+	if len(signersList) == 0 {
+		activeAccount := [][]string{[]string{"no-signers"}}
+		signersList = append(signersList, activeAccount...)
+	}
 
-	for i, argumentList := range argumentLists {
-		formattedTitle := title
-		if len(argumentList) > 0 {
-			formattedTitle = fmt.Sprintf(
-				"%s %s %s",
-				title,
-				argumentListConjunction,
-				entryPointInfo.pragmaArgumentStrings[i],
-			)
-		}
+	for index, argumentList := range argumentLists {
```

Comment:
This loop is very complex. Try to break it into smaller functions that are easier to reason about